Feature/check who is loser

choo121600 opened the pull request:
## Motivation ⍰

- 벌칙자 기능 구현 및 데이터베이스에 put


<br>

## Key Changes 🔑


### 승패자 여부와 상관 없이 select 한 것들 백엔드로 전달
#### 그를 위해 selectFirstQuestionPlayers, selectSecondQuestionPlayers를 만들어 출력시 필요한 name과 index를 넣어줌
https://github.com/Dream-Catch/Balance-Catch-iOS/blob/914d65d975f6158a51a0cdb392490b42a8b76c6f/Balance-Catch-iOS/Balance-Catch-iOS/View/WhoIsLoserView.swift#L102-L110

#### 백엔드로 전달
https://github.com/Dream-Catch/Balance-Catch-iOS/blob/914d65d975f6158a51a0cdb392490b42a8b76c6f/Balance-Catch-iOS/Balance-Catch-iOS/View/WhoIsLoserView.swift#L95-L99

#### 패자들을 loserList에 넣어줌
https://github.com/Dream-Catch/Balance-Catch-iOS/blob/914d65d975f6158a51a0cdb392490b42a8b76c6f/Balance-Catch-iOS/Balance-Catch-iOS/View/WhoIsLoserView.swift#L112-L129


<br>

## To Reviewers 🙏🏻

- [x] 모두 당첨 됐을 때, 모두 당첨 안 됐을 때, 5명이상 당첨됐을 때 체크해주세요

<br>

## Linked Issue 🔗

- [x] #59

choo121600 committed:
Feat: 패자 기능 구현

choo121600 committed:
Fix: loserQuestionIdx 수정

choo121600 committed:
Merge branch 'develop' into Feature/CheckWhoIsLoser

choo121600 committed:
Fix: put 처리 지움

choo121600 commented:
Put처리하는 것을 RecommandOrNotView에서 처리함으로서 불필요한 리스트(selectFirstQuestionPlayers, selectSecondQuestionPlayers) 생성을 막고, 불필요한 함수(setPlayerSelecedtList, putQuestionScore) 지움.
또한, setLoserPlayerList함수 수정

```swift
    private func setLoserPlayerList () {
        for (index, player) in playerList.players.enumerated() {
            if player.select == loserQuestionIdx {
                loserList.append((name: player.name, index: index))
            } else if loserQuestionIdx == -1 {
                loserList.append((name: player.name, index: index))
            }
        }
    }
```

Minny27 committed:
[#59]Feat: 벌칙 당첨자 뷰 커스텀
- 임의가 걸렸을 때, 다 걸렸을 때, 아무도 안 결렸을 떄 세 가지 케이스로 나눔
- 이전에 구현되어 있던 ViewHeightKey로 scroll view 동적 높이 구현
- 최대 벌칙 당첨자가 4.5명 보이도록 설정
- 스크롤하게 끔 유도하기 위해서 .5 높이를 추가
- PlayerModel에서 select를 가독성을 위해서 SelectType으로 수정
- SelectType이 .none이면 draw를 처리하기 위함
- RecommandOrNotView에서 Skip버튼 간격 디자인 반영
- info.plist 파일 위치 변경

Minny27 commented:
다음을 수정했어요
1. 벌칙 당첨자 스크롤 뷰의 동적 높이 구현

이전에 구현되어 있던 PreferenceKey 프로토콜을 이용해서 동적 높이 구현
코드는 다음과 같음
```swift
struct ViewHeightKey: PreferenceKey {
    static var defaultValue: CGFloat = 0
    static var maxValue: CGFloat = 0 // 이 값을 onAppear()에서 수정
    
    static func reduce(value: inout CGFloat, nextValue: () -> CGFloat) {
        let newValue = max(value, nextValue())
        value = maxValue >= newValue ? newValue : maxValue
    }
}
```

maxValue가 static변수이기 때문에 값의 일관성을 유지할 수 없어서 onAppear()에서 값을 초기화해줌
그리고 max 높이는 최대 4.5개가 보이도록 설정
그 이유는 iphone se 3세대를 최소 디바이스로 가정했을 때 4개까지가 적당했습니다.
코드는 다음과 같음
```swift
// WhoIsLoserView.swift
.onAppear {
    ViewHeightKey.maxValue = CGFloat.loserViewHeight * 4 + CGFloat.loserViewHeight / 2 // 벌칙자가 최대 3개까지 보이도록 설정
}
```
스크린샷은 다음과 같음
<img src="https://github.com/Dream-Catch/Balance-Catch-iOS/assets/68800789/f909f750-36d5-4b15-9ecb-178793caf97b" width=25%>

2. 3가지 케이스를 나눠서 뷰 표시

다 걸렸을 때, 모두 안 결렸을 때, 누군가 걸렸을 때 세 가지 케이스로 나눔
```swift
var body: some View {
      ZStack {
          VStack(alignment: .center) {
              // 다 걸렸을 때
              if loserList.count == playerList.players.count || loserSelectType == .none {
                  Text("모두 벌칙에 당첨됐습니다!")
                      .font(.title)
                      .fontWeight(.bold)
                      .padding(.bottom, 14)
              }
              
              // 아무도 안 걸렸을 때
              else if loserList.isEmpty {
                  Text("다들 심리전에 고수시군요~")
                      .font(.title)
                      .fontWeight(.bold)
                      .padding(.bottom, 14)
              }
              
              // 누군가 걸렸을 때
              else {
                  Text("벌칙 당첨자💣")
                      .font(.system(size: 36, weight: .bold))
                      .shadow(color:.gray,radius:2,x:3,y:3)
                      .padding(.bottom, 46)
{ ... }

```

이 때, SelectType으로 enum을 만들어서 loserSelectType이 first, second, none 케이스로 나뉘고 각 케이스가 진 질문의 케이스입니다. 여기서 none은 draw를 처리하기 위함

3. playerModel 수정

index 0, 1, -1보다 가독성을 높이고 싶었음
코드는 다음과 같음
```swift
//  PlayerModel.swift
import Foundation

enum SelectType {
    case first, second, none
}

struct Player {
    var index: Int
    var name: String
    var selectType: SelectType
}

class PlayerList: ObservableObject {
    @Published var players: [Player]
    
    init(players: [Player] = []) {
        self.players = players
    }
    
    func addPlayer(index: Int,
                   name: String) {
        let player = Player(index: index,
                            name: name,
                            selectType : .none)
        players.append(player)
    }
    
    func resetPlayers() {
        players.removeAll()
    }
}

```

Minny27 committed:
[#59] CGFloat에 벌칙자 뷰 높이 추가

Minny27 committed:
[#59] 마지막 화면에서 벌칙 당첨 혹은 당첨되지 않았을 때의 뷰 커스텀
- LoserButtonStyle 커스텀
- 텍스트에 Button Style이 먹지 않아서 Button에 텍스트 넣고 disable하는 방식으로 구현
- GameResult라는 enum은 원래 대중픽 결과에서 사용하는 열거형인데 이 열거형을 공유
- svg 이미지 추가 -> 화질 더 좋음
- size 및 이미지 문자열 추가

Minny27 commented:
https://github.com/Dream-Catch/Balance-Catch-iOS/blob/914d65d975f6158a51a0cdb392490b42a8b76c6f/Balance-Catch-iOS/Balance-Catch-iOS/View/WhoIsLoserView.swift#L102-L110

Minny27 committed:
[#59] 그림자 수정
- 아래쪽만 그림자 적용되도록 수정
- blur 4로 수정

Minny27 commented:
잘 작동되서 머지 할게요!
